README
=head1 NAME

MojoX::Tusu - Apache-like dispatcher for Mojolicious

=head1 SYNOPSIS

    sub startup {
        ....

        use MojoX::Tusu;
        my $tusu = MojoX::Tusu->new($self);
        
        # following is optional
        
        $tusu->document_root($self->home->rel_dir('www2'));
        $tusu->extensions_to_render([qw(html htm xml txt)]);
        
        # initialize Text::PSTemplate::Plugable if necessary
        
        my $plugin_instance = $tusu->plug('some_plugin');
        $tusu->engine->set_...();
        $tusu->plug('Your::Component', 'YC');
        
        $r->route('/specific/path')->to(cb => sub {
            $tusu->bootstrap($_[0], 'Your::Component', 'your_method');
        });
    }

=head1 DESCRIPTION

C<MojoX::Tusu> is a sub framework on Mojolicious using Text::PSTemplate
for renderer. With this framework, you can deploy directory based web sites
onto Mojolicious at once.

This framework automatically activate own dispacher which behaves like apache
web server. You can build your web site into single document root directory
named public_html in hierarchal structure. The document root directory can
contain both server-parsed-documents and static files such as images.

MojoX::Tusu doesn't require files to be named like index.html.ep style but just
like index.html. You can specify which files to be server parsable by telling
it extensions. It also provides some more apache-like features such as
directory_index, error_document and file permissions checking.

One of the intent of this module is to enhance existing static websites into
dynamic with minimal effort. The chances are that most typical website data are
transplantable with no change at all.

=head2 Components & Plugins

Mojo::Tusu provides object oriented component & plugin framework. You can
easily add your custom features into your website. The following is an example
for plugin development.

    <span><% questionize('Hello') %></span>

To make it possible, you should write a module like this. 

    package MyUtility;
    use strict;
    use warnings;
    use base 'MojoX::Tusu::PluginBase';
    
    sub questionize : TplExport {
        my ($self, $sentence) = @_;
        my $c = $self->controller; # mojolicious controller in case you need
        return $sentence . '?';
    }

To activate this plugin, you must plug-in this at mojolicious startup method.

    sub startup {
        my $self = shift;
        my $tusu = MojoX::Tusu->new($self);
        $tusu->plug('YourUtility', ''); ## second argument is the namespace
    }

The following is an example for component development.

    <div id="productContainer">
        <% Product::list_by_category('books', 10) %>
    </div>

To make it possible, you should write a module like this.

    package Product;
    use strict;
    use warnings;
    use base 'MojoX::Tusu::ComponentBase';
    
    sub init {
        my ($self, $app) = @_;
        $self->set_ini({...}); ### DB SETTING OR SOMETHING
    }
    
    sub list_by_category : TplExport {
        my ($self, $category, $limit) = @_;
        my $c = $self->controller; # mojolicious controller in case you need
        
        # MAY BE ACCESS TO YOUR MODELS HERE
        
        return $html_snippet;
    }

To activate this component, you must plug-in this at mojolicious startup method.

    sub startup {
        my $self = shift;
        my $tusu = MojoX::Tusu->new($self);
        $tusu->plug('Product');
    }

The only difference between plugins and components is that components can have
an init method to have own data.

=head1 METHODS

=head2 MojoX::Tusu->new($app)

Constractor. This method takes Mojolicious app for argument and returns
MojoX::Tusu instance.
    
    $tusu = MojoX::Tusu->new($app)

=head2 $instance->document_root($directory)

This method sets root directory for templates and static files. Following
example is default setting.
    
    my $tusu = MojoX::Tusu->new($app);
    $tusu->document_root($self->home->rel_dir('public_html'));

=head2 directory_index($candidate1 [, $candidate2])

This method sets default file names for searching files in directory when
the request path doesn't ended with file name. And this setting also affects to
inside template context such as include('path') function. Following example is
the default setting.

    my $tusu = MojoX::Tusu->new($app);
    $tusu->directory_index(['index.html', 'index.htm']);

=head2 $instance->engine

This returns Text::PSTemplate::Plugable instance. You can customize the template
system behavior by calling parser methods directly.
    
    my $tusu = MojoX::Tusu->new($app);
    my $pst = $tusu->engine;
    $pst->set_delimiter('<!--', '-->');

=head2 $instance->plug($plug_name, [$namespace])

This is a delegate method for Text::PSTemplate::Plugable->plug just for hooking
init method for components. All arguments are thrown at
Text::PSTemplate::Plugable->plug.

    my $tusu = MojoX::Tusu->new($self);
    $tusu->plug('Text::PSTemplate::Plugin::HTML', 'HTML');

You also can plug multiple plugins at once.

    my $tusu = MojoX::Tusu->new($self);
    $tusu->plug(
        'Namespace::A' => 'A',   # namespace is A
        'Namespace::B' => '',    # namespace is ''
        'Namespace::C' => undef, # namespace is Namespace::C
    );

=head2 $instance->bootstrap($controller, $component, $method)

This method is a sub dispacher method. You can specify a class and a method the
route to be dispached to.

    my $tusu = MojoX::Tusu->new($self);
    $r->route('/some/path')->via('post')->to(cb => sub {
        $tusu->bootstrap($c, 'Your::Component', 'post');
    });

=head2 $instance->extensions_to_render($array_ref)

This method sets the extensions to be parsed by tusu renderer. If request
doesn't match any of extensions, dispatcher try to render it as static file.
Following settting is the default.

    my $tusu = MojoX::Tusu->new($self);
    $tusu->extensions_to_render(['html','htm','xml'])

=head2 $instance->error_document($hash_ref)

This method setup custom error pages like apache's ErrorDocument.

    $instance->error_document({
        404 => '/errors/404.html',
        403 => '/errors/403.html',
        500 => '/errors/405.html',
    })

=head1 SEE ALSO

L<Mojolicious>, L<Text::PSTemplate>, L<MojoX::Renderer>

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sugama Keita.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


Changes
Revision history for Mojolicious-Plugin-Tusu

0.39 2012/02/17

0.38 2012/02/17
  - Fixed a bug on handling multibyte file names indexing.

0.37 2012/02/17
  - Fixed a bug on installation.

0.36 2012/02/16
  - Added indexes option for generating file list for directory access.

0.35 2012/02/15
  - Changed internal codes in favor of Mojolicious v2.49

0.34 2012/02/08
  - Bundled Text::PSTemplate ver0.44 to fix error message for undefined funcs
  - Stoped bundling Mojolicious.

0.33 2012/01/18
  - Added user_err function for templates.
  - Removed put_user_err function for templates.
  - Improved component function for templates.

0.32 2012/01/17
  - Added user_err method for components and put_user_err for templates.

0.31 2012/01/12
  - Added component method for template.

0.30 2012/01/11
  - Added redirect_to method.

0.29 2011/12/28
  - Added get_component method.
  - Added render method on ComponentBase
  - Bundled Text::PSTemplate v0.41

0.28 2011/11/29
  - Generator namespaces has changed
  - Fixed a bug on generator that some file's permission was wrong

0.26 2011/11/25
  - M::T::ComponentBase::ini and set_ini removed. Use Mojo::Base::attr instead.
  - Text::PSTemplate v0.40 bundled.
  - Removed run_hook_reverse call in favor of emit_hook_reverse
  
0.25 2011/07/14
  - Tusu now is a Mojolicious plugin
  - PluginBase removed. ComponentBase is available instead.
  - Text::PSTemplate v0.36 bundled.

0.24 2011/07/14
  - Short cut for html escape activated

0.23 2011/07/09
  - Dependency on Mojolicious updated
  - A dependency for test specified

0.22 2011/06/30
  - output_encoding feature no longer available. Consider using
    Plack::Middleware instead.
  - Permission check is eased to inside document root

0.21 2011/06/22
  - ini and ini_set added in ComponentBase
  - Depencency updated
  
0.20 2011/06/21
  - Mojolicious::Command::Generate::TusuApp added
  - Static file bug on win fixed
  
0.19 2011/06/20
  - Pathes in templates without trailing slash are no longer autofilled with
    directory index
  - FIX: CGI mode bug on root access
  - input & output encoding can be auto detected

0.18 2011/06/10
  - url_abs added
  - Permission check activated
  - Directory Index can be a static file now
  - Fixed a bug on CGI mode
  - pst handler removed and tusu is available
  - error document method added

0.17 2011/06/08
  - File searching rule has been changed. It got similar behavior to apache 
  - directory_index method added
  - Compatibility with apache improved
  - bootstrap renewed

0.16 2011/06/01
  - Reduce messing app attr namaspace

0.15 2011/06/01
  - plug method now capable of multiple plugins at once

0.14 2011/06/01
  - document_root method optimized & bug fixed

0.13 2011/05/31
  - Extensions_to_render method added

0.12 2011/05/26
  - Exception object explicitly stringified

0.11 2011/05/23
  - Template dir and static dir both now defaults to public_html

0.10 2011/05/20
  - Mojo::Tusu::plug for hooking MojoX::Tusu::ComponentBase::init

0.09 2011/05/19
  - Distribution renamed

0.08 2011/05/19
  - Controller no longer given via second arg on tpl funcs

0.07 2011/05/18
  - ComponentBase got some utility functions for template

0.06 2011/05/18
  - API got simple

0.05 2011/05/17
  - Template not found now causes die

0.04 2011/05/16
  - ActionBase class added for frameworking

0.03 2011/04/22
  - Improved error logging

0.02 2011/03/31
  - some natie helpers available

0.01 2011/03/23
  - initial release


lib/Mojolicious/Plugin/Tusu.pm
package Mojolicious::Plugin::Tusu;
use strict;
use warnings;
use Try::Tiny;
use Text::PSTemplate;
use Mojo::Base 'Mojolicious::Plugin';
use Mojo::Util;
use Carp;
use File::Basename;
use File::Spec;
use Mojolicious::Types;

    our $APP;
    our $CONTROLLER;
    
    __PACKAGE__->attr('engine');
    __PACKAGE__->attr('extensions_to_render');
    __PACKAGE__->attr('directory_index');
    __PACKAGE__->attr('error_document');
    __PACKAGE__->attr('document_root');
    __PACKAGE__->attr('indexes');
    
    # internal use
    __PACKAGE__->attr('_default_route_set');
    
    sub _check_hash_key {
        my ($hash_ref, @keys) = @_;
        my %keys;
        $keys{$_} = 1 foreach(@keys);
        for my $key (keys %$hash_ref) {
            if (! $keys{$key}) {
                return $key;
            }
        }
    }

    sub register {
        my ($self, $app, $args) = @_;
        
        my $default_args = {
            document_root           => $app->home->rel_dir('public_html'),
            encoding                => 'utf8',
            extensions_to_render    => ['html','htm','xml'],
            directory_index         => ['index.html','index.htm'],
            error_document          => {},
            components              => {},
            indexes                 => 0,
        };
        
        if (my $key = _check_hash_key($args, keys %$default_args)) {
            croak "Unknown argument $key";
        }
        
        $args = {%$default_args, %$args};
        
        if (! -d $args->{document_root}) {
            die qq{Document root "$args->{document_root}" is not a directory};
        }
        
        $app->hook(after_build_tx => sub {
            my $app = $_[1];
            if (! $self->_default_route_set) {
                $self->_default_route_set(1);
                $app->routes
                    ->route('/:template', template => qr{.*})
                    ->name('')
                    ->to(cb => sub {$_[0]->render(handler => 'tusu')});
            }
        });
        
        $app->hook('around_dispatch' => sub {
            my ($next, $c) = @_;
            $self->_dispatch($c->app, $c);
            $next->();
        });
        
        $app->static->paths([$args->{document_root}, $self->_asset]);
        $app->renderer->paths([$args->{document_root}]);
        
        my $engine = Text::PSTemplate->new;
        $engine->set_filter('=', \&Mojo::Util::html_escape);
        $engine->set_filename_trans_coderef(sub {
            _filename_trans($args->{document_root}, $args->{directory_index}, @_);
        });
        
        {
            local $APP = $app;
            $engine->plug(
                'MojoX::Tusu::ComponentBase'            => undef,
                'MojoX::Tusu::Component::Util'          => '',
                'MojoX::Tusu::Component::Mojolicious'   => 'Mojolicious',
                %{$args->{components}}
            );
        }
        
        $engine->set_encoding($args->{encoding});
        
        $self->engine($engine);
        $self->directory_index($args->{directory_index});
        $self->error_document($args->{error_document});
        $self->extensions_to_render($args->{extensions_to_render});
        $self->document_root($args->{document_root});
        $self->indexes($args->{indexes});
        
        $app->renderer->add_handler(tusu => sub { $self->_render(@_) });
        
        return $self;
    }
    
    sub get_component {
        my ($self, $name) = @_;
        return $self->engine->get_plugin($name);
    }
    
    ### ---
    ### bootstrap for frameworking
    ### ---
    sub bootstrap {
        my ($self, $c, $component, $action, @args) = @_;
        local $CONTROLLER = $c;
        return $self->engine->get_plugin($component)->$action(@args);
    }
    
    ### ---
    ### Custom dispatcher
    ### ---
    sub _dispatch {
        my ($self, $app, $c) = @_;
        
        my $tx = $c->tx;
        if ($tx->is_websocket) {
            $c->res->code(undef);
        }
        $app->sessions->load($c);
        my $plugins = $app->plugins;
        $plugins->emit_hook(before_dispatch => $c);
        
        my $path = $tx->req->url->path->to_string || '/';
        
        my $not_found;
        
        my $check_result = $self->_check_file_type($path);
        
        if (! $check_result->{type}) {
            if ($path =~ qr{/$} && $self->indexes) {
                $self->_render_indexes($c);
                return;
            } else {
                $not_found = 1;
            }
        } elsif ($check_result->{type} eq 'directory') {
            $c->redirect_to($path. '/');
            $tx->res->code(301);
            return;
        } elsif (! _permission_ok($check_result->{path}, $app->static->paths->[0])) {
            $self->_render_error_document($c, 403);
            return;
        }
        
        if (! $not_found) {
            
            my $path = $check_result->{path};
            
            ### dynamic content
            for my $ext (@{$self->extensions_to_render}) {
                if ($path !~ m{\.} || $path =~ m{\.$ext$}) {
                    my $res = $tx->res;
                    if (my $code = ($tx->req->error)[1]) {
                        $res->code($code)
                    } elsif ($tx->is_websocket) {
                        $res->code(426)
                    }
                    if (! $app->routes->dispatch($c) || ! $res->code) {
                        $c->render_not_found;
                    }
                    return;
                }
            }
            
            ## This must not be happen
            if ($path =~ m{((\.(cgi|php|rb))|/)$}) {
                $self->_render_error_document($c, 403);
                return;
            }
        }
        
        my $relpath =
            ($check_result->{path})
                ? File::Spec->abs2rel($check_result->{path}, $app->static->paths->[0])
                : $path;
        ### defaults to static content
        if ($app->static->serve($c, $relpath)) {
            $c->stash->{'mojo.static'} = 1;
            $c->rendered;
        }
        if (! $tx->res->code) {
            $self->_render_error_document($c, 404);
        }
        $plugins->emit_hook_reverse(after_static_dispatch => $c);
    }

    ### ---
    ### Asset directory
    ### ---
    sub _asset {
        my ($class, $path) = @_;
        if ($path) {
            return File::Spec->catfile(dirname(__FILE__), 'Tusu', 'Asset', $path);
        }
        return File::Spec->catdir(dirname(__FILE__), 'Tusu', 'Asset');
    }
    
    sub _file_to_mime_class {
        my $name = shift;
        my $ext = ($name =~ qr{\.(\w+)$}) ? $1 : '';
        return (split('/', Mojolicious::Types->type($ext) || 'text/plain'))[0];
    }
    
    sub _file_timestamp {
        my $path = shift;
        my @dt = localtime((stat($path))[9]);
        return sprintf('%d-%02d-%02d %02d:%02d', 1900 + $dt[5], $dt[4] + 1, $dt[3], $dt[2], $dt[1]);
    }
    
    sub _file_size {
        my $path = shift;
        return ((stat($path))[7] > 1024)
            ? sprintf("%.1f",(stat($path))[7] / 1024) . 'KB'
            : (stat($path))[7]. 'B';
    }
    
    ### ---
    ### Render file list
    ### ---
    sub _render_indexes {
        my ($self, $c) = @_;
        my $req_path = $c->tx->req->url->path;
        my $dir = $self->document_root() . $req_path;
        
        opendir(my $DIR, $dir);
        my @file = readdir($DIR);
        closedir $DIR;
        
        my @dataset = ();
        for my $file (@file) {
            utf8::decode($file);
            if ($file =~ qr{^\.$} || $req_path =~ qr{^/$} && $file =~ qr{^\.\.$}) {
                next;
            }
            my $path = $dir. $file;
            push(@dataset, {
                name        => -f $path ? $file : $file. '/',
                timestamp   => _file_timestamp($path),
                size        => _file_size($path),
                type        => -f $path ? _file_to_mime_class($file) : 'dir',
            });
        }
        @dataset = sort {
            ($a->{type} ne 'dir') <=> ($b->{type} ne 'dir')
            ||
            $a->{name} cmp $b->{name}
        } @dataset;
        my $tpl = Text::PSTemplate->new;
        $tpl->set_var(dir => $req_path, dataset => \@dataset);
        $c->render_text($tpl->parse_file($self->_asset('file_list.html')));
    }
    
    ### ---
    ### Render Error document
    ### ---
    sub _render_error_document {
        my ($self, $c, $code, $debug_message) = @_;
        
        $debug_message ||= 'Unknown Error';
        
        $c->app->log->debug($debug_message);
        
        if ($c->app->mode eq 'production') {
            if (my $template = $self->error_document->{$code}) {
                $c->render(handler => 'tusu', template => $template);
                $c->rendered($code);
                return;
            }
        }
        if ($code == 404) {
            $c->render_not_found;
        } else {
            $c->render_exception($debug_message);
        }
        $c->res->code($code);
    }
    
    ### ---
    ### fill directory_index candidate
    ### ---
    sub _fill_filename {
        my ($path, $directory_index) = @_;
        for my $default (@{$directory_index}) {
            my $path = File::Spec->catfile($path, $default);
            if (-f $path) {
                return $path;
            }
        }
        return;
    }
    
    ### ---
    ### find file and type
    ### ---
    sub _check_file_type {
        my ($self, $name) = @_;
        $name ||= '';
        my $leading_slash  = (substr($name, 0, 1) eq '/');
        my $trailing_slash = (substr($name, -1, 1) eq '/');
        $name =~ s{^/}{};
        my $path = File::Spec->catfile($self->document_root, $name);
        if (-f $path) {
            return {type => 'file', path => $path};
        }
        if ($trailing_slash) {
            if (my $fixed_path = _fill_filename($path, $self->directory_index)) {
                return {type => 'file', path => $fixed_path};
            }
        } elsif (-d $path) {
            return {type => 'directory'};
        }
        return {};
    }
    
    ### ---
    ### foo/bar.html    -> public_html/foo/bar.html
    ### foo/            -> public_html/foo/index.html
    ### foo             -> public_html/foo
    ### ---
    sub _filename_trans {
        my ($template_base, $directory_index, $name) = @_;
        $name ||= '';
        my $leading_slash = substr($name, 0, 1) eq '/';
        my $trailing_slash = substr($name, -1, 1) eq '/';
        $name =~ s{^/}{};
        my $dir;
        if ($leading_slash) {
            $dir = $template_base;
        } else {
            $dir = (File::Spec->splitpath(Text::PSTemplate->get_current_filename))[1];
        }
        my $path = File::Spec->catfile($dir, $name);
        if ($trailing_slash) {
            if (my $fixed_path = _fill_filename($path, $directory_index)) {
                return $fixed_path;
            }
        }
        return $path;
    }
    
    ### ---
    ### Check if others readable
    ### ---
    sub _permission_ok {
        my ($name, $base) = @_;
        $base ||= '';
        if ($^O eq 'MSWin32') {
            return 1;
        }
        if ($name && -f $name && ((stat($name))[2] & 4)) {
            $name =~ s{(^|/)[^/]+$}{};
            while (-d $name) {
                if ($name eq $base) {
                    return 1;
                }
                if (! ((stat($name))[2] & 1)) {
                    return 0;
                }
                $name =~ s{(^|/)[^/]+$}{};
            }
            return 1;
        }
        return 0;
    }
    
    ### ---
    ### tusu renderer
    ### ---
    sub _render {
        my ($self, $renderer, $c, $output, $options) = @_;
        
        try {
            local $SIG{__DIE__} = undef;
            local $CONTROLLER = $c;
            $$output = Text::PSTemplate ->new($self->engine)
                                        ->parse_file('/'. $options->{template});
        } catch {
            my $err = $_ || 'Unknown Error';
            $c->app->log->error(qq(Template error in "$options->{template}": $err));
            $self->_render_error_document($c, 500, "$err");
            $$output = '';
            return 0;
        };
        return 1;
    }

1;

__END__

=head1 NAME

Mojolicious::Plugin::Tusu - Apache-like dispatcher for Mojolicious

=head1 SYNOPSIS

    use Mojolicious::Plugin::Tusu;

For non lite app

    sub startup {
        my $self = shift;
        my $tusu = $self->plugin(tusu => {});
    }

OR

    sub startup {
        my $self = shift;
        my $tusu = $self->plugin(tusu => {
            document_root => $self->home->rel_dir('www2'),
            components => {
                'Your::Component' => 'YC',
            },
            extensions_to_render => [qw(html htm xml txt)],
        });
        
        $r->route('/specific/path')->to(cb => sub {
            $tusu->bootstrap($_[0], 'Your::Component', 'your_method');
        });
    }

For lite app

    my $tusu = plugin tusu => {...};

=head1 DESCRIPTION

C<Mojolicious::Plugin::Tusu> is a sub framework on Mojolicious using
Text::PSTemplate for renderer. With this framework, you can deploy directory
based web sites onto Mojolicious at once.

This framework automatically activate own dispatcher which behaves like apache
web server. You can build your web site into single document root directory
named public_html in hierarchal structure. The document root directory can
contain both server-parsed-documents and static files such as images.

Mojolicious::Plugin::Tusu doesn't require files to be named like index.html.ep
style but just like index.html. You can specify which files to be server
parsable by telling it the extensions. It also provides some more apache-like
features such as directory_index, error_document and file permissions checking.

One of the intent of this module is to enhance existing static websites into
dynamic with minimal effort. The chances are that most typical website data are
transplantable with no change at all.

=head1 OPTIONS

=head2 document_root => string

This option sets root directory for templates and static files. Following
example is default setting.

    my $tusu = $self->plugin(tusu => {
        document_root => $self->home->rel_dir('public_html')
    });

=head2 components => hash

    my $tusu = $self->plugin(tusu => {
        components => {
            'Namespace::A' => 'A',   # namespace is A
            'Namespace::B' => '',    # namespace is ''
            'Namespace::C' => undef, # namespace is Namespace::C
        },
    });

=head2 encoding => string or array ref

This option sets encoding for template files. Array ref causes auto detection
active.

    my $tusu = $self->plugin(tusu => {
        encoding => 'Shift-JIS',
    });
    
    or..
    
    my $tusu = $self->plugin(tusu => {
        encoding => ['Shift-JIS', 'utf8'],
    });

=head2 directory_index => array ref

This option sets default file names for searching files in directory when
the request path doesn't ended with file name. And this setting also affects to
inside template context such as include('path') function. Following example is
the default setting.

    my $tusu = $self->plugin(tusu => {
        directory_index => ['index.html', 'index.htm'],
    });

=head2 indexes => bool

This option emulates apache's indexes option. When the value is 1,
the server generates file list page for directory access.

    my $tusu = $self->plugin(tusu => {
        indexes => 1,
    });

=head2 extensions_to_render => array ref

This option sets the extensions to be parsed by tusu renderer. If request
doesn't match any of extensions, dispatcher try to render it as static file.
Following setting is the default.

    my $tusu = $self->plugin(tusu => {
        extensions_to_render => ['html','htm','xml'],
    });

=head2 error_document => hash ref

This option setup custom error pages like apache's ErrorDocument.

    my $tusu = $self->plugin(tusu => {
        error_document => {
            404 => '/errors/404.html',
            403 => '/errors/403.html',
            500 => '/errors/405.html',
        },
    });

=head1 METHODS

=head2 Mojolicious::Plugin::Tusu->new($app)

Constructor. 
    
    $tusu = Mojolicious::Plugin::Tusu->new($app)

=head2 $instance->register($app)

This method internally called.

=head2 $instance->engine

This returns Text::PSTemplate instance. You can customize the template system
behavior by calling parser methods directly.
    
    my $tusu = Mojolicious::Plugin::Tusu->new($app);
    my $pst = $tusu->engine;
    $pst->set_delimiter('<!--', '-->');

=head2 $instance->bootstrap($controller, $component, $method)

This method is a sub dispatcher method. You can specify a class and a method the
route to be dispatched to.

    my $tusu = Mojolicious::Plugin::Tusu->new($self);
    $r->route('/some/path')->via('post')->to(cb => sub {
        $tusu->bootstrap($c, 'Your::Component', 'post');
    });

=head2 $instance->get_component($name)

This is an alias to Text::PSTemplate->get_plugin. With this method, you can get
component instance.

    my $component = $tusu->get_component('Your::Component');

=head1 What does Tusu mean?

Tusu means mojo in Ainu languages which is spoken by the native inhabitants of
Hokkaido prefecture, Japan.

=head1 SEE ALSO

L<Mojolicious>, L<Text::PSTemplate>

L<http://en.wikipedia.org/wiki/Ainu_languages>

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sugama Keita.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
